add initial block formatter implementation

haxscramper opened the pull request:
Added implementation of the text layouter for error message or code formatter. Split from the #224

This PR provides test and implementation of the generic text block layout algorithm, which can be used for error message formatting and general code layout. Split from aforementioned documentation generator PR was done in order to simplify latter one and because formatter implementation is usable on its own (specifically needed for #332).

Reimplements https://github.com/google/rfmt

haxscramper commented:
The PR implementation is ready for the review, I will add test now.

saem commented on lib/experimental/dod_helpers.nim:
```diff
@@ -0,0 +1,184 @@
+import std/[macros, strutils, options]
```
```suggestion
## Contains various types, procs and templates for data oriented design style
## programming/memory layouts. See: https://www.dataorienteddesign.com/dodbook/

import std/[macros, strutils, options]
```

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1884 @@
+## https://github.com/google/rfmt.
+
+# Thanks to nim being syntactically close to python this is mostly
+# just blatant copy-paste of python code with added type annotations.
```
```suggestion
## Nimskull reimplementation of layout algorithm from
## https://github.com/google/rfmt.

# Thanks to nimskull being syntactically close to python this is mostly
# just blatant copy-paste of python code with added type annotations.
```

easier when we need to do a big ol' rename.

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1884 @@
+]
+
+const
+  infty = 1024 * 1024 * 1024 * 1024
```
```suggestion
  infty = 1024 * 1024 * 1024 * 1024
    ## info type bit mask thingy??
```

not sure what it should say yet, but that was my first guess besides "infinity"

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1884 @@
+#*************************************************************************#
+#===========================  Type definition  ===========================#
+
+import ./dod_helpers
```
boo, relative imports!

(I'm guessing it's due to the main module scenario)

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+  LayoutElement = ref object
+    ## An element of a layout object - a directive to the console.
+    ##
+    ## This class sports a collection of static methods, each of which
```
```suggestion
    ## This object sports a collection of static methods, each of which
```

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+    ## returns an anonymous function invoking a method of the console
+    ## to which it is applied.
+    ##
+    ## Refer to the corresponding methods of the LytConsole class for
```
```suggestion
    ## Refer to the corresponding methods of the LytConsole object for
```

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+
+    id {.requiresInit.}: int
+
+    knots: seq[int] ## a list of ints, specifying the margin settings at
+    ## which the layout changes. Note that the first knot is required to be
+    ## 0.
+    spans: seq[int] ## a list of ints, giving for each knot, the width of
+    ## the corresponding layout in characters.
+    intercepts: seq[float] ## constant cost associated with each knot -
+    ## computed cost of the outputting this layout solution at each know in
+    ## the `knots` list
+    gradients: seq[float] ## at each knot, the rate with which the layout
```
Might be worth making distincts?

canelhasmateus commented on tests/stdlib/ttext_layouter.nim:
```diff
@@ -0,0 +1,205 @@
+
+let comm = s.str(", ")
+
+echo toString(
```
Aside from the ones inside `for`, are these `echo` ammenable to being transformed into `doAssert`?

canelhasmateus replied:
oh, not easily. Just noticed the toString definition above.

haxscramper replied:
Yes, will do now

haxscramper replied:
moved all checks into doAssert

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+    knots: seq[int] ## a list of ints, specifying the margin settings at
+    ## which the layout changes. Note that the first knot is required to be
+    ## 0.
+    spans: seq[int] ## a list of ints, giving for each knot, the width of
```
is this meant to be one entry in `knots` will always have an entry in `spans` or is the relationship not strict 1-to-1?

haxscramper replied:
one to one

saem replied:
Follow-up: can one have many things refer to the exact same solution?

It's a ref type which is why I'm asking.

saem replied:
need to understand the transform internals better, again.

deferred.

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+    len: int
+
+  LytBlock* = ref object
+    layoutCache: Table[Option[LytSolution], Option[LytSolution]]
```
wait... even the key has an option?

as in we can have a single nil key with a possibly nil (optional) layout?

saem replied:
need to understand the black box better.

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1751 @@
+        discard
+
+
+  Layout* = ref object
```
Can many `LayoutElements` refer to the same `Layout`?

saem replied:
this is because of python porting/blackbox.

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1754 @@
+declareHighMasking(LytStr, highMaskRange = 8)
+
+
+type
```
A preamble here might help, this is me making a guess, but I'm just trying to draw a "map" of
relationship between the types.

```suggestion
type
  ## These are types for formatting text. Text formatting doesn't have a perfect solution that works
  ## in all cases and we have to resort to heuristics to choose the most approriate text layout on a
  ## case by case basis.
  ##
  ## Text is initially described via a DSL as a set of blocks (`LytBlock`). When it's time to output the
  ## text a layout (`LytLayout`) determines the exact strings (`LytStr`) to output. Text layout requires
  ## heuristics to choose between various options we use a solver ('LytSolution') which is used to
  ## determine the precise layout choosen for a given block.
```

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1754 @@
+    strs: seq[LytStr]
+    len: int
+
+  LytBlock* = ref object
```
do we expect to have have many things refer to the same block?

haxscramper replied:
I don't understand the question. A layout block is a single block in the text layout. It does not refer to anything nor does anyone refer to it, aside from parent combinator blocks like line/stack/combinator/choice. If you are asking why the type is a `ref` - because it forms a tree that is supposed to be randomly added to or added into another trees, similar to how AST is constructed.

saem replied:
yup, it was because it's a ref.

saem replied:
OK, this is a look at separately/later if the value types make sense.

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1754 @@
+    len: int
+
+  LytBlock* = ref object
+    layoutCache: Table[Option[LytSolution], Option[LytSolution]]
```
If we can pop the `layoutCache` out, by having a block id and a separate data structure to store the cache itself, then we could have the DSL defined in one chunk/module and the layout solving/pipeline entirely separately.

If the break-up works then someone can read the DSL part alone for just consuming/using the API, might be easier to learn it over time.

saem replied:
for later refactoring.

saem reviewed (commented):
remarked on a few data type related matters, need to look at the algorithms.

I think it might be very helpful if we can use the DOD bits to split:
* dsl and types to build blocks
* transformer/solver
* the very simple output bits that can consume from the previous bit

saem commented on lib/experimental/text_layouter.nim:
```diff
@@ -0,0 +1,1762 @@
+
+proc minSolution(solutions: seq[LytSolution]): Option[LytSolution] =
+  ## Form the piecewise minimum of a sequence of LytSolutions.
+
```
observation:

I find these visual breaks have me jumping back up to ensure I suddenly didn't lose context -- destroys my skim speed.

It's not a problem when I'm reading carefully, but quickly skimming is really hard because I'm purposefully not reading everything carefully, so I suddenly doubt things and wonder if I missed something and now I've fallen out of skim mode and am back tracking.

haxscramper replied:
I don't understand what "visual breaks" do you refer to. Do you mean the `#=====================  LytSolution manipulation logic  =====================#` comment above the line you commented on?

saem replied:
```nim
  ## Form the piecewise minimum of a sequence of LytSolutions.
  # (blank line is the visual break)
  ## more docs
```

The blank line in the block doc comment, I left it as an observation.

It's merely a data point so over some period of time (bigger than this PR) we eventually figure out a style guide.

haxscramper replied:
fixed to continuous comments now

haxscramper committed:
generic text layout algorithm implementation

add implementation of the generic text layout algorithm for use in error
message formatting, code beautifier and more. Implementation itself is
based on the https://github.com/google/rfmt. This commit does not change
any other parts of the codebase aside from introducing the layouter.

saem commented:
Small thing would be add some explanation around motivation to the PR message body, but either way I'm ok to bors.

haxscramper commented:
updated the description

haxscramper commented:
bors r+
